internal/controller/compute/statefulserverset/datavolume_controller.go

```
@@ -0,0 +1,165 @@
+		return v1alpha1.Volume{}, err
+	}
+	k.log.Info("Finished creating Data Volume", "name", name)
+
+	return *kubeVolume, nil
+}
+
+// Get - returns a volume kubernetes object
+func (k *kubeDataVolumeController) Get(ctx context.Context, volumeName, ns string) (*v1alpha1.Volume, error) {
+	obj := &v1alpha1.Volume{}
+	err := k.kube.Get(ctx, types.NamespacedName{
+		Namespace: ns,
+		Name:      volumeName,
+	}, obj)
+	return obj, err
+}
+
+// Delete - deletes the datavolume k8s client and waits until it is deleted
+func (k *kubeDataVolumeController) Delete(ctx context.Context, name, namespace string) error {
+	condemnedVolume, err := k.Get(ctx, name, namespace)
+	if err != nil {
+		return err
+	}
+	if err := k.kube.Delete(ctx, condemnedVolume); err != nil {
+		return fmt.Errorf("error deleting data volume %w", err)
+	}
+	return serverset.WaitForKubeResource(ctx, serverset.ResourceReadyTimeout, k.isDataVolumeDeleted, condemnedVolume.Name, namespace)
+}
+
+// isAvailable - checks if a volume is available
+func (k *kubeDataVolumeController) isAvailable(ctx context.Context, name, namespace string) (bool, error) {
+	obj, err := k.Get(ctx, name, namespace)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	if obj != nil && obj.Status.AtProvider.VolumeID != "" && strings.EqualFold(obj.Status.AtProvider.State, ionoscloud.Available) {
+		return true, nil
+	}
+	return false, err
+}
+
+func (k *kubeDataVolumeController) isDataVolumeDeleted(ctx context.Context, name, namespace string) (bool, error) {
+	k.log.Info("Checking if data volume is deleted", "name", name, "namespace", namespace)
+	obj := &v1alpha1.Volume{}
+	err := k.kube.Get(ctx, types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}, obj)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			k.log.Info("Data volume has been deleted", "name", name, "namespace", namespace)
+			return true, nil
+		}
+		return false, nil
+	}
+	return false, nil
```

We could write it like this to get rid of nested ifs:
```go
	if err != nil && apiErrors.IsNotFound(err) {
		k.log.Info("Data volume has been deleted", "name", name, "namespace", namespace)
		return true, nil
	} else if err != nil && !apiErrors.IsNotFound(err) {
		return false, nil
	}
	return false, nil
```

---

This might actually be a small issue since err is lost if it's != notfound. have to check though

---

we decided to go with what was originally written